elasticsearch/src/main/scala/com/velocidi/apso/elasticsearch/ElasticsearchBulkInserter.scala

```diff
@@ -85,20 +85,19 @@ class ElasticsearchBulkInserter(
     }
   }
 
-  private[this] def countSuccessfulMsgs(sentBuffer: Iterable[Message], bulkResponse: BulkResponse) = {
+  // returns a list of the failed requests
+  private[this] def notifyOfSuccessfulAndGetFailed(sentBuffer: Iterable[Message], bulkResponse: BulkResponse) = {
     val reqsAndRes = sentBuffer.iterator.zip(bulkResponse.items.iterator)
 
-    val (successCount, failedReqs) = reqsAndRes.foldLeft((0, List.empty[Message])) {
-      case ((count, failedReqs), (req, res)) if res.error.isDefined =>
-        (count, failedReqs ::: addRetry(req, res))
+    reqsAndRes.foldLeft(List.empty[Message]) {
```

Comment:
I think this can now be replaced by a `flatMap`, which should be more efficient (since iternally the list concatenation will use a mutable builder). I think we can also get rid of the guard clause, which should avoid calling `unapply` twice on the happy path.

I haven't tested it, but I think something like this should work:

```scala
reqsAndRes.flatMap { case (req, res) =>
  if (res.error.isDefined) {
    addRetry(req, res)
  }
  else {
    req.sender ! res.id 
    tryCountMap.remove(req) 
    Nil
  }
}
```

Reply:
Yes, that seems to make sense. Thank you! Improved in e546c98